Save screenshots when end2end tests fail

toofar opened the pull request:
* Takes a screen capture of the XVFB screen when an e2e test fails
* It won't take any if you aren't using xvfb
* On local runs it'll put the screenshots in $TEMP/pytest-current/pytest-screenshots/
* On GitHub actions it'll attach any screenshots as run artifacts in a zip file
* Handles multiple test workers running in parallel (eg they don't clobber the output dir)
* Only captures the frontmost window if multiple are open
* Will capture screenshots for tests that are rerun by https://github.com/pytest-dev/pytest-rerunfailures, which is not ideal because we don't get error logs for the failed-by-rerun tests

Example: <https://github.com/qutebrowser/qutebrowser/actions/runs/10537503218?pr=8243#:~:text=Produced%20during%20runtime>

![image](https://github.com/user-attachments/assets/538dd656-12bb-49af-a0b0-076d73a0eabb)

The-Compiler commented:
I haven't had a chance to take a closer look yet, but have you considered using [tmp_path](https://docs.pytest.org/en/stable/how-to/tmp_path.html) instead of rolling your own file handling, locking, filename sanitization etc. etc.? It's made for things like this, as pytest keeps the last 3 temporary paths around even after a test finishes.

toofar commented:
Okay, ready for review again, changes I've made:

* moved screenshots to be saved in the pytest temp dir
* made the CI runners use RUNNER_TEMP as the temp dir instead of /tmp so upload-artifacts can find the pytest temp dir
* made the screenshot filenames simpler so we don't need any extra sanitation
* added a line to the pytest output (right above the short test summary) pointing users to the directory with screenshots saved in it

toofar commented:
I'll merge this in a couple of days, I've already used it a couple of times from the branch.

toofar commented:
Locally I'm seeing errors like this during teardown sometimes.

![image](https://github.com/user-attachments/assets/0e38bd6e-05d1-4584-897c-4c724b582aca)

Potentially it's due to using the take_x11_screenshot fixture from after_test, which looks like it's during teardown? Not 100% sure what's going on there because that fixture should have been called earlier in the request already. Could be related to https://github.com/pytest-dev/pytest/issues/9984, but I don't know why the assertion error has so much stuff in it.
Not sure what to do about that apart from move the screenshot stuff back to not being in a pytest fixture.

The-Compiler commented:
Hmmm, I think on top of whatever is going wrong, something inside pytest decides the innermost part of the traceback is uninteresting and [hides it](https://docs.pytest.org/en/stable/example/simple.html#writing-well-integrated-assertion-helpers) (possibly because someone assumed this assertion would never be triggered?).

Can you try getting it fail with `--tb=native`? While that won't give use the information what value triggered the `AssertionError` anymore (the long line from your screenshot), it hopefully will give us the exact location...

toofar commented:
Ah, thank you. That does show where the assertion is.

<details>
<summary>Here's a stack trace with `--tb=native` as is</summary>

```
________ ERROR at teardown of test_sandboxing[disable-seccomp-bpf-True-False-True-You are NOT adequately sandboxed.] _________
Traceback (most recent call last):
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 341, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 242, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/threadexception.py", line 97, in pytest_runtest_teardown
    yield from thread_exception_runtest_hook()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
    yield
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 100, in pytest_runtest_teardown
    yield from unraisable_exception_runtest_hook()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
    yield
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/logging.py", line 853, in pytest_runtest_teardown
    yield from self._runtest_for(item, "teardown")
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/logging.py", line 829, in _runtest_for
    yield
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/capture.py", line 885, in pytest_runtest_teardown
    return (yield)
            ^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/pytestqt/plugin.py", line 207, in pytest_runtest_teardown
    result = yield
             ^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 189, in pytest_runtest_teardown
    item.session._setupstate.teardown_exact(nextitem)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 557, in teardown_exact
    raise exceptions[0]
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 546, in teardown_exact
    fin()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1032, in finish
    raise exceptions[0]
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1021, in finish
    fin()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 907, in _teardown_yield_fixture
    next(it)
  File "/work/tests/end2end/fixtures/quteprocess.py", line 984, in quteproc_new
    proc.after_test()
  File "/work/tests/end2end/fixtures/quteprocess.py", line 552, in after_test
    self._take_x11_screenshot_of_failed_test()
  File "/work/tests/end2end/fixtures/quteprocess.py", line 881, in _take_x11_screenshot_of_failed_test
    fixture = self.request.getfixturevalue('take_x11_screenshot')
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 532, in getfixturevalue
    fixturedef = self._get_active_fixturedef(argname)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 617, in _get_active_fixturedef
    fixturedef.execute(request=subrequest)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1047, in execute
    fixturedef = request._get_active_fixturedef(argname)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 617, in _get_active_fixturedef
    fixturedef.execute(request=subrequest)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1094, in execute
    request.node.addfinalizer(finalizer)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/nodes.py", line 391, in addfinalizer
    self.session._setupstate.addfinalizer(fin, self)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 526, in addfinalizer
    assert node in self.stack, (node, self.stack)
           ^^^^^^^^^^^^^^^^^^
AssertionError: (<Function test_sandboxing[disable-seccomp-bpf-True-False-True-You are NOT adequately sandboxed.]>, {<Session  exitstatus=<ExitCode.OK: 0> testsfailed=3 testscollected=3>: ([<bound method Node.teardown of <Session  exitstatus=<ExitCode.OK: 0> testsfailed=3 testscollected=3>>, functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='check_display' scope='session' baseid='tests'>>, request=<SubRequest 'check_display' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='check_yaml_c_exts' scope='session' baseid='tests'>>, request=<SubRequest 'check_yaml_c_exts' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='fail_on_logging' scope='session' baseid='tests'>>, request=<SubRequest 'fail_on_logging' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='qapp_args' scope='session' baseid='tests'>>, request=<SubRequest 'qapp_args' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='qapp_cls' scope='session' baseid=''>>, request=<SubRequest 'qapp_cls' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='pytestconfig' scope='session' baseid=''>>, request=<SubRequest 'pytestconfig' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='qapp' scope='session' baseid=''>>, request=<SubRequest 'qapp' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='qapp' scope='session' baseid='tests'>>, request=<SubRequest 'qapp' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='server' scope='session' baseid='tests/end2end'>>, request=<SubRequest 'server' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>), functools.partial(<bound method FixtureDef.finish of <FixtureDef argname='tmp_path_factory' scope='session' baseid=''>>, request=<SubRequest 'tmp_path_factory' for <Function test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]>>)], None), <Dir work>: ([<bound method Node.teardown of <Dir work>>], None), <Dir tests>: ([<bound method Node.teardown of <Dir tests>>], None), <Dir end2end>: ([<bound method Node.teardown of <Dir end2end>>], None), <Module test_invocations.py>: ([<bound method Node.teardown of <Module test_invocations.py>>], None)})
```
</details>

<details>
<summary>And one with the parameterize mark removed, just in case it was that (doesn't look like it)</summary>

```
Traceback (most recent call last):
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 341, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 242, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/threadexception.py", line 97, in pytest_runtest_teardown
    yield from thread_exception_runtest_hook()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
    yield
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 100, in pytest_runtest_teardown
    yield from unraisable_exception_runtest_hook()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
    yield
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/logging.py", line 853, in pytest_runtest_teardown
    yield from self._runtest_for(item, "teardown")
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/logging.py", line 829, in _runtest_for
    yield
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/capture.py", line 885, in pytest_runtest_teardown
    return (yield)
            ^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/pytestqt/plugin.py", line 207, in pytest_runtest_teardown
    result = yield
             ^^^^^
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 189, in pytest_runtest_teardown
    item.session._setupstate.teardown_exact(nextitem)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 557, in teardown_exact
    raise exceptions[0]
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 546, in teardown_exact
    fin()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1032, in finish
    raise exceptions[0]
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1021, in finish
    fin()
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 907, in _teardown_yield_fixture
    next(it)
  File "/work/tests/end2end/fixtures/quteprocess.py", line 984, in quteproc_new
    proc.after_test()
  File "/work/tests/end2end/fixtures/quteprocess.py", line 552, in after_test
    self._take_x11_screenshot_of_failed_test()
  File "/work/tests/end2end/fixtures/quteprocess.py", line 881, in _take_x11_screenshot_of_failed_test
    fixture = self.request.getfixturevalue('take_x11_screenshot')
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 532, in getfixturevalue
    fixturedef = self._get_active_fixturedef(argname)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 617, in _get_active_fixturedef
    fixturedef.execute(request=subrequest)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1047, in execute
    fixturedef = request._get_active_fixturedef(argname)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 617, in _get_active_fixturedef
    fixturedef.execute(request=subrequest)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/fixtures.py", line 1094, in execute
    request.node.addfinalizer(finalizer)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/nodes.py", line 391, in addfinalizer
    self.session._setupstate.addfinalizer(fin, self)
  File "/home/user/.tox/py/lib/python3.12/site-packages/_pytest/runner.py", line 526, in addfinalizer
    assert node in self.stack, (node, self.stack)
           ^^^^^^^^^^^^^^^^^^
AssertionError: (<Function test_sandboxing>, {})
```
</details>

This is with `tox -e py -- tests/end2end/test_invocations.py -k sandboxing -s --tb=native` within a `archlinux-webengine-unstable-qt6` docker container. Haven't tried comparing it to other configurations yet.

I'll probably just ignore this for a week, or two, to give me time to focus on the 6.8 permissions stuff which is higher priority (and also seeing [weird failures](https://github.com/qutebrowser/qutebrowser/tree/feat/68_permissions_askeverytime), surely unrelated). We can revert this if it causes issues and I'll add it back even better later :)

The-Compiler commented:
Weird. Opened #8333 for this, but seems like potentially a pytest bug.

webknjaz commented on doc/changelog.asciidoc:
```diff
@@ -30,6 +30,10 @@ Removed
 
 - Support for macOS 11 Big Sur is dropped. Binaries are now built on macOS 12
   Monterey and are unlikely to still run on older macOS versions.
+- Failed end2end tests will now save screenshots of the browser window when
```
@The-Compiler should this be under the "Removed" section? It seems like it might be misplaced, looking at the changes in the PR.

The-Compiler replied:
That indeed seems like a mistake, well spotted! Pushed a commit to fix it.
